### ISSUE TITLE
IEEE 754 Negative zeros get no respect. 

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#3977]
##### Presenting problem:

The following code succeeds on JVM (22 for the greater problem on which I am working) and fails
with Scala Native 0.5.5-SNAPSHOT.

```
   assertTrue(
      "-0.0 is not strictly less than 0.0",
      (new jl.Double(-0.0)).compareTo(new jl.Double(0.0)) < 0
    )

```

I would be astonished if Float does not have the same problem.

##### Background:
There appears to be no SN Test case for `compareTo`.  There also appears to be no
Test from Scala.js which might be ported.

##### Hasty & terse analysis:

This is probably because SN `Double#compareTo` delegates directly to `Double#compare`.
Java 22, and probably earlier, is quite explicit and unusually clear about the former comparing
negative zero less than positive zero.  The doc on the latter says nothing, but JVM seems to 
follow the same comparison order.  

Note that various equals methods will correctly compare the two as the same.

```
   assertEquals(
      "Fi", 66,
      jl.Double.compare(-0.0, 0.0)
    )
```

The `Double.compare()` code is pretty dense. I think it has to be looked at from the 
perspective of either argument being -0.0 (and the other +0.0).  

I think IEEE 754 double format has a bit which distinguishes negative & positive zero.
Trick is to add the test without adding execution cost to the most frequently executed paths.

Suspects, because appear to be only places where I can get the 0 comparison result I am seeing, rather
than the -1 JVM returns:
```
// JVM returns 'false' for this because of the numeric comparison != to the right of the 'and'.
// I suspect but have not checked at runtime that SN does the same.
//

 else if (x == y && 0.0d != x) 0  // LeeT, I suspect this has a bug when y is -0.0
```
Java `Double.equals()` is documented to distinguish from  -0.0 & 0.0 (and return false)
```
scala> (-0.0).equals(0.0)
res8: Boolean = false

// Problematic numeric comparisons
scala> -0.0 == 0.0
res9: Boolean = true

scala> -0.0 < 0.0
res10: Boolean = false
```

Looks like this is doing an IEEE 754 check on the sign (highest, leftmost) bit.
_Should_ distinguish negative and positive zero if code ever gets here.

I would have to trace, but it seems to need a zero check (and a few comments).
If something other than zeros get here, the lower 63 bits are discarded.

```
 } else {
        val d1 = doubleToRawLongBits(x)
	val d2 = doubleToRawLongBits(y)
        ((d1 >> 63) - (d2 >> 63)).toInt
      }
```

**Comment by LeeTibbert** [comment]
I am about 4 or 5 recursion levels deep into pending fixes, so I can not address this in the short run (days).
I'll comment it out the check which revealed this to make progress in my current work and hope to circle
back to this issue.

**Comment by LeeTibbert** [comment]
Quick update.

The code I most suspect now is `(new jl.Double(-0.0)`, especially how the value gets
passed to the constructor. It appears that the compare/compareTo code is, to a 
first approximation, fine in SN (I have private Scala.js tests running in my sandbox and relevant tests pass).

`-0.0.compareTo(new jl.Double(0.0))` works as expected on both JVM & SN.
`(new jl.Double(-0.0)).compareTo(0.0))` Passes on JVM and fails on SN.

I have a number of other passing and failing cases which I will try to organize for report.
Basically both the String and Double overloads of `Double.valueOf` work on both JVM & SN.
Since the String `valueOf` calls `parseDouble`, the latter is probably good.

On the other hand, output methods report that the Double object from `(new jl.Double(-0.0)`
is +0.0, not the expected/intended/required/correct -0.0.  `toDouble` and `doubleValue` 
are in the category.  

I need to do _much_ more tracing before I believe any of this.

**Comment by LeeTibbert** [comment]
Superseded by PR #3982, which has a more minimal reproducer and shorter discussion trail.
